### PR TITLE
Add registry authentication in runtime

### PR DESCRIPTION
**Pull request by kvega005**
## PR Checklist

- [ ] **Closes:** Link to issue #xxx
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated if needed and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated if needed
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/wsl/) and link it here: #xxx

### CODE REVIEW

File: src/windows/wslcsession/WSLCSession.cpp [hunk 1]
```diff
@@ -1273,6 +1291,32 @@ try
...
+    RETURN_HR_IF_NULL(E_POINTER, Image);
+
+    auto lock = m_lock.lock_shared();
+    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), !m_dockerClient.has_value());
+
+    auto [repo, tagOrDigest] = wslutil::ParseImage(Image);
+
+    std::optional<std::string> registryAuth;
+
+    if (RegistryAuthenticationInformation != nullptr && *RegistryAuthenticationInformation != '\0')
```
[the remark this answers is not in the document]
Comment (kvega005) [reply src/windows/wslcsession/WSLCSession.cpp]: When omitted, Docker pushes all tags for that image, not just latest. So the reviewer's concern about it failing is wrong — it won't fail, it'll actually push all tags.

File: src/windows/wslcsession/DockerHTTPClient.cpp [hunk 2]
```diff
@@ -650,11 +690,16 @@ std::unique_ptr<DockerHTTPClient::HTTPRequestContext> DockerHTTPClient::SendRequ
...
+    for (const auto& [field, value] : Headers)
     {
         req.set(field, value);
     }
 
+    for (const auto& [name, value] : CustomHeaders)
+    {
+        req.set(name, value);
+    }
+
```
[the remark this answers is not in the document]
Comment (kvega005) [reply src/windows/wslcsession/DockerHTTPClient.cpp]: The header value is forwarded to dockerd over a local Unix domain socket inside the VM. The caller is the session owner who already has full control over the session. I don't think additional validation is warranted here, but happy to add a length check for consistency if preferred.

File: src/windows/service/inc/wslc.idl [hunk 3]
```diff
@@ -591,7 +591,9 @@ interface IWSLCSession : IUnknown
     HRESULT ListImages([in, unique] const WSLCListImageOptions* Options, [out, size_is(, *Count)] WSLCImageInformation** Images, [out] ULONG* Count);
     HRESULT DeleteImage([in] const WSLCDeleteImageOptions* Options, [out, size_is(, *Count)] WSLCDeletedImageInformation** DeletedImages, [out] ULONG* Count);
     HRESULT TagImage([in] const WSLCTagImageOptions* Options);
+    HRESULT PushImage([in] LPCSTR Image, [in, unique] LPCSTR RegistryAuthenticationInformation, [in, unique] IProgressCallback* ProgressCallback);
     HRESULT InspectImage([in] LPCSTR ImageNameOrId, [out] LPSTR* Output);
+    HRESULT Authenticate([in] LPCSTR ServerAddress, [in] LPCSTR Username, [in] LPCSTR Password, [out] LPSTR* IdentityToken);
```
[the remark this answers is not in the document]
Comment (kvega005) [reply src/windows/service/inc/wslc.idl]: @OneBlue, @benhillis do we need a `IWSLCSession2` ? Or is this not required while we are still in preview?

[the remark this answers is not in the document]
Comment (benhillis) [reply src/windows/service/inc/wslc.idl]: In general it’s better to add com methods to the end, but we can still change the interface during preview.

File: src/windows/common/WslcCredentialStore.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+Abstract:
+
+    Implementation of credential store helpers.
+
+--*/
+
+#include "WslcCredentialStore.h"
+#include <nlohmann/json.hpp>
+#include <wincrypt.h>
+
```
[the remark this answers is not in the document]
Comment (benhillis) [reply src/windows/common/WslcCredentialStore.cpp]: These are valid

File: src/windows/service/inc/wslc.idl [hunk 5]
```diff
@@ -614,7 +614,9 @@ interface IWSLCSession : IUnknown
     HRESULT ListImages([in, unique] const WSLCListImageOptions* Options, [out, size_is(, *Count)] WSLCImageInformation** Images, [out] ULONG* Count);
     HRESULT DeleteImage([in] const WSLCDeleteImageOptions* Options, [out, size_is(, *Count)] WSLCDeletedImageInformation** DeletedImages, [out] ULONG* Count);
     HRESULT TagImage([in] const WSLCTagImageOptions* Options);
+    HRESULT PushImage([in] LPCSTR Image, [in] LPCSTR RegistryAuthenticationInformation, [in, unique] IProgressCallback* ProgressCallback);
     HRESULT InspectImage([in] LPCSTR ImageNameOrId, [out] LPSTR* Output);
```
[the remark this answers is not in the document]
Comment (benhillis) [reply src/windows/service/inc/wslc.idl:637]: Can the sdk be udated as part of this change?

File: src/windows/service/inc/wslc.idl [hunk 6]
```diff
@@ -614,7 +614,9 @@ interface IWSLCSession : IUnknown
     HRESULT ListImages([in, unique] const WSLCListImageOptions* Options, [out, size_is(, *Count)] WSLCImageInformation** Images, [out] ULONG* Count);
     HRESULT DeleteImage([in] const WSLCDeleteImageOptions* Options, [out, size_is(, *Count)] WSLCDeletedImageInformation** DeletedImages, [out] ULONG* Count);
     HRESULT TagImage([in] const WSLCTagImageOptions* Options);
+    HRESULT PushImage([in] LPCSTR Image, [in] LPCSTR RegistryAuthenticationInformation, [in, unique] IProgressCallback* ProgressCallback);
```
[the remark this answers is not in the document]
Comment (kvega005) [reply src/windows/service/inc/wslc.idl]: Docker requires auth token for Push. Authorization token is optional for pull.

File: src/windows/service/inc/wslc.idl (hunk 5 above)
[the remark this answers is not in the document]
Comment (kvega005) [reply src/windows/service/inc/wslc.idl:637]: I was planning to update the SDK and CLI in a follow up PR. But the SDK changes on their own are pretty small, so I can definately add it here.

File: test/windows/WSLCTests.cpp [hunk 7]
```diff
@@ -473,74 +513,93 @@ class WSLCTests
...
+    {
+        WSL2_TEST_ONLY();
+
+        constexpr auto c_username = "wslctest";
+        constexpr auto c_password = "password";
+
+        auto registry = wsl::windows::common::WSLCLocalRegistry::Start(*m_defaultSession, c_username, c_password);
+
+        auto registryUrl = std::format(L"http://{}/v2/", registry.GetServerAddress());
+        auto registryAddress = registry.GetServerAddress();
```
[the remark this answers is not in the document]
Comment (kvega005) [reply test/windows/WSLCTests.cpp]: Not a compile error. Works as intended

File: src/windows/service/inc/wslc.idl (hunk 5 above)
[the remark this answers is not in the document]
Comment (kvega005) [reply src/windows/service/inc/wslc.idl:637]: The WslcRegistryAuthenticationInformation struct in the SDK header was a TBD placeholder — it's an empty struct and the SDK implementation has always passed nullptr for the auth parameter to PullImage. Since it was never wired up or functional, I think it's safe to change now.

My plan is to:

1. Replace the `WslcRegistryAuthenticationInformation` struct / `authInfo` pointer in `WslcPullImageOptions` with a `PCSTR registryAuth` field (matching what the runtime actually expects — a base64-encoded auth string)
2. Add `WslcAuthenticate()` so SDK consumers can authenticate with a registry and get a token
3. Add `WslcPushSessionImage()` and `WslcTagSessionImage()` to complete the push flow

If we're worried about source compatibility for anyone already referencing the empty struct, I could alternatively keep the old WslcPullSessionImage signature as-is (it never worked for auth anyway) and add a new WslcPullSessionImageWithAuth that takes the string. But given the struct was literally // TBD with no members and no functional path, I'd prefer just fixing it in place. Thoughts?

File: src/windows/common/WslcCredentialStore.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+--*/
+
+#include "precomp.h"
+#include "WslcCredentialStore.h"
+#include <wincrypt.h>
+
+namespace {
+
+std::string Base64Encode(const std::string& input)
```
Comment (OneBlue) [inline src/windows/common/WslcCredentialStore.cpp]: nit: We should factor this with the same method in WSLCSession.cpp

File: src/windows/common/WSLCLocalRegistry.cpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
+/*++
+
```
Comment (OneBlue) [inline src/windows/common/WSLCLocalRegistry.cpp]: nit: I would recommend moving this file to `test/windows` since this is only used in tests

File: src/windows/wslcsession/DockerHTTPClient.cpp [hunk 10]
```diff
@@ -632,7 +662,11 @@ void DockerHTTPClient::DockerHttpResponseHandle::OnResponseBytes(const gsl::span
 }
 
 std::unique_ptr<DockerHTTPClient::HTTPRequestContext> DockerHTTPClient::SendRequestImpl(
-    verb Method, const URL& Url, const std::string& Body, const std::map<boost::beast::http::field, std::string>& Headers)
+    verb Method,
+    const URL& Url,
+    const std::string& Body,
+    const std::map<boost::beast::http::field, std::string>& Headers,
+    const std::map<std::string, std::string>& CustomHeaders)
```
Comment (OneBlue) [inline src/windows/wslcsession/DockerHTTPClient.cpp]: If we need to pass headers that boost doesn't support, I'd say let's merge them all into a `std::map<std::string, std::string>` so we only have one argument for headers

File: test/windows/WSLCTests.cpp [hunk 11]
```diff
@@ -85,6 +87,21 @@ class WSLCTests
...
+        }
+
+        if (!hasImage("alpine:latest"))
+        {
+            LoadTestImage("alpine:latest");
+        }
+
+        if (!hasImage("wslc-registry:latest"))
+        {
+            LoadTestImage("wslc-registry:latest");
```
Comment (OneBlue) [inline test/windows/WSLCTests.cpp:100]: Could we check-in a .ps1 / dockerfile so we easily rebuild this image in the future if we need to ?

File: src/windows/common/WSLCLocalRegistry.cpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    m_password(std::move(password)),
+    m_serverAddress(std::move(serverAddress)),
+    m_container(std::move(container))
+{
+}
+
+WSLCLocalRegistry::~WSLCLocalRegistry()
+{
+    // Delete the container first while the session is still active.
+    m_container.Reset();
```
Comment (OneBlue) [inline src/windows/common/WSLCLocalRegistry.cpp]: nit: This doesn't need to be done explicitly

**Review by OneBlue — COMMENTED** [review]
LGTM, minor comments

File: src/windows/service/inc/wslc.idl [hunk 13]
```diff
@@ -656,6 +656,9 @@ interface IWSLCSession : IUnknown
     HRESULT DeleteVolume([in] LPCSTR Name);
     HRESULT ListVolumes([out, size_is(, *Count)] WSLCVolumeInformation** Volumes, [out] ULONG* Count);
     HRESULT InspectVolume([in] LPCSTR Name, [out] LPSTR* Output);
+
+    HRESULT Authenticate([in] LPCSTR ServerAddress, [in] LPCSTR Username, [in] LPCSTR Password, [out] LPSTR* IdentityToken);
+    HRESULT PushImage([in] LPCSTR Image, [in] LPCSTR RegistryAuthenticationInformation, [in, unique] IProgressCallback* ProgressCallback);
```
Comment (yao-msft) [inline src/windows/service/inc/wslc.idl:682]: Is RegistryAuthenticationInformation the IdentityToken returned in Authenticate()? If they are the same, I think it'd be better to use same name.
And nit: PushImage should be placed around line 617 to be together with other image management methods.

Comment (kvega005) [reply src/windows/service/inc/wslc.idl:682]: It's not the same. `RegistryAuthenticationInformation` is a base64 encode JSON string that contains the server address and credentials. It can be the token, or it can be the username and password. Not all registry servers support tokens.

File: src/windows/service/inc/wslc.idl (hunk 5 above)
[the remark this answers is not in the document]
Comment (OneBlue) [reply src/windows/service/inc/wslc.idl:637]: > * Replace the `WslcRegistryAuthenticationInformation` struct / `authInfo` pointer in `WslcPullImageOptions` with a `PCSTR registryAuth` field (matching what the runtime actually expects — a base64-encoded auth string)
> * Add `WslcAuthenticate()` so SDK consumers can authenticate with a registry and get a token
> * Add `WslcPushSessionImage()` and `WslcTagSessionImage()` to complete the push flow

I think this is the right call. I wouldn't worry about source compatibility, especially given that the previous structure didn't do anything, I wouldn't expect anyone to have taken a dependency on it. 

Regarding the SDK changes, if it's just a matter of plumbing the calls through to the service, I'd say let's include it in this change if possible, otherwise if it's a bigger change, it's probably worth its own PR

File: tools/test/images/wslc-registry/entrypoint.sh [hunk 14]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
+set -e
+
+if [ -n "$USERNAME" ]; then
+    mkdir -p /auth
+    htpasswd -Bbn "$USERNAME" "$PASSWORD" > /auth/htpasswd
+
+    export REGISTRY_AUTH=htpasswd
+    export REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd
+    export REGISTRY_AUTH_HTPASSWD_REALM="WSLC Registry"
```
[the remark this answers is not in the document]
Comment (kvega005) [reply tools/test/images/wslc-registry/entrypoint.sh:10]: This is fine. Its just used for testing.

File: test/windows/WslcSdkTests.cpp [hunk 15]
```diff
@@ -2053,6 +1979,288 @@ class WslcSdkTests
...
+
+        // Get the IWSLCSession COM object from the SDK session handle.
+        auto& session = *reinterpret_cast<WslcSessionImpl*>(m_defaultSession)->session;
+        auto container = launcher.Launch(session, WSLCContainerStartFlagsNone);
+
+        auto registryAddress = std::format("127.0.0.1:{}", port);
+
+        // Wait for the registry to be ready by probing from the host.
+        auto hostUrl = std::format(L"http://{}", registryAddress);
+        ExpectHttpResponse(hostUrl.c_str(), 200, true);
```
[the remark this answers is not in the document]
Comment (kvega005) [reply test/windows/WslcSdkTests.cpp:1942]: This test is using a known registry version, so probing `http://127.0.0.1:<port>` is deterministic

File: test/windows/WslcSdkTests.cpp [hunk 16]
```diff
@@ -2053,6 +1979,288 @@ class WslcSdkTests
...
+    // -----------------------------------------------------------------------
+
+    // Starts a local registry container with host-mode networking and returns [container, registryAddress].
+    // Uses the COM API (via GetInternalType) with WSLCContainerLauncher to get host-mode networking,
+    // which the SDK doesn't expose. Host networking shares the VM's network namespace, so the registry
+    // is reachable at 127.0.0.1:<port> from both dockerd (inside the VM) and the host.
+    std::pair<wsl::windows::common::RunningWSLCContainer, std::string> StartLocalRegistry(
+        const std::string& username = {}, const std::string& password = {}, uint16_t port = 5000)
+    {
+        std::vector<std::string> env = {std::format("REGISTRY_HTTP_ADDR=0.0.0.0:{}", port)};
```
[the remark this answers is not in the document]
Comment (kvega005) [reply test/windows/WslcSdkTests.cpp:1923]: I kept 5000 for simplicity, but we can default to a high port number, maybe something in the ephemeral port range. Although I don't really think we need to know for now.

[the remark this answers is not in the document]
Comment (OneBlue) [reply test/windows/WslcSdkTests.cpp:1923]: This is probably something that we can improve once we have proper support for ephemeral port binds. Definitely not a blocker for this PR

File: test/windows/WSLCTests.cpp [hunk 17]
```diff
@@ -473,74 +529,91 @@ class WSLCTests
...
-                VERIFY_FAIL();
-            }
-            else
-            {
-                ExpectImagePresent(*m_defaultSession, ExpectedTag->c_str());
-            }
+            ExpectImagePresent(*m_defaultSession, registryImage.c_str());
         };
 
-        validatePull("ubuntu@sha256:2e863c44b718727c860746568e1d54afd13b2fa71b160f5cd9058fc436217b30", {});
```
Comment (OneBlue) [inline test/windows/WSLCTests.cpp:516]: nit: We might want to keep the other images refs in a test case that's disabled for now, those were useful to debug corner cases with PullImages()

**Review by OneBlue — APPROVED** [review]
LGTM

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 18]
```diff
@@ -1282,6 +1280,72 @@ try
 }
 CATCH_RETURN();
 
+STDAPI WslcTagSessionImage(_In_ WslcSession session, _In_ const WslcTagImageOptions* options, _Outptr_opt_result_z_ PWSTR* errorMessage)
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp:1283]: @JohnMcPMS to review the SDK changes

File: src/windows/wslcsession/WSLCSession.cpp [hunk 19]
```diff
@@ -1274,6 +1264,25 @@ try
...
 CATCH_RETURN();
 
+HRESULT WSLCSession::PushImage(LPCSTR Image, LPCSTR RegistryAuthenticationInformation, IProgressCallback* ProgressCallback)
+try
+{
+    COMServiceExecutionContext context;
+
+    RETURN_HR_IF_NULL(E_POINTER, Image);
+    RETURN_HR_IF_NULL(E_POINTER, RegistryAuthenticationInformation);
+
```
[the remark this answers is not in the document]
Comment (kvega005) [reply src/windows/wslcsession/WSLCSession.cpp:1274]: Docker validates the string. We dont need to.

**Review by OneBlue — APPROVED** [review]

File: src/windows/WslcSDK/wslcsdk.h [hunk 20]
```diff
@@ -457,6 +452,58 @@ typedef struct WslcImageInfo
...
+    _In_z_ PCSTR repo;  // Target repository name.
+    _In_z_ PCSTR tag;   // Target tag name.
+} WslcTagImageOptions;
+
+STDAPI WslcTagSessionImage(_In_ WslcSession session, _In_ const WslcTagImageOptions* options, _Outptr_opt_result_z_ PWSTR* errorMessage);
+
+typedef struct WslcPushImageOptions
+{
+    _In_z_ PCSTR image;
+    _In_z_ PCSTR registryAuth; // Base64-encoded X-Registry-Auth header value.
```
Comment (JohnMcPMS) [inline src/windows/WslcSDK/wslcsdk.h:467]: Should this be optional?  I realize that it is unlikely to be optional in the real world, but it doesn't seem like a strict requirement for functionality.

File: src/windows/wslcsession/DockerHTTPClient.cpp [hunk 21]
```diff
@@ -162,16 +163,20 @@ std::unique_ptr<DockerHTTPClient::HTTPRequestContext> DockerHTTPClient::PullImag
...
 }
 
 std::unique_ptr<DockerHTTPClient::HTTPRequestContext> DockerHTTPClient::LoadImage(uint64_t ContentLength)
 {
     return SendRequestImpl(
-        verb::post,
-        URL::Create("/images/load"),
-        {},
-        {{http::field::content_type, "application/x-tar"}, {http::field::content_length, std::to_string(ContentLength)}});
+        verb::post, URL::Create("/images/load"), {}, {{"Content-Type", "application/x-tar"}, {"Content-Length", std::to_string(ContentLength)}});
```
Comment (JohnMcPMS) [inline src/windows/wslcsession/DockerHTTPClient.cpp:179]: We should have our own list of header name definitions to prevent typos.

File: src/windows/wslcsession/DockerHTTPClient.cpp [hunk 22]
```diff
@@ -194,6 +198,28 @@ void DockerHTTPClient::TagImage(const std::string& Id, const std::string& Repo,
...
+    const std::string& ImageName, const std::optional<std::string>& tag, const std::string& registryAuth)
+{
+    auto url = URL::Create("/images/{}/push", ImageName);
+
+    if (tag.has_value())
+    {
+        url.SetParameter("tag", tag.value());
+    }
+
+    std::map<std::string, std::string> customHeaders = {{"X-Registry-Auth", registryAuth}};
```
Comment (JohnMcPMS) [inline src/windows/wslcsession/DockerHTTPClient.cpp:211]: Second use of `X-Registry-Auth`, consider defining a symbol for shared use.

**Review by JohnMcPMS — APPROVED** [review]
Minor comments from my side.

File: src/windows/WslcSDK/wslcsdk.h (hunk 20 above)
Comment (kvega005) [reply src/windows/WslcSDK/wslcsdk.h:467]: It's not optional in the docker daemon, which is why I did not make it optional here. But we could add some code in the runtime or SDK to fill it with an empty Base64 encoded Json auth string.

Comment (kvega005) [reply src/windows/WslcSDK/wslcsdk.h:467]: If we pass an empty string or NULL to the docker daemon, docker will throw an error.

File: src/windows/wslcsession/DockerHTTPClient.cpp (hunk 21 above)
Comment (kvega005) [reply src/windows/wslcsession/DockerHTTPClient.cpp:179]: Makes sense. I will follow up on this in another PR.
